### ISSUE TITLE
Make SubCapture thread safe

### ISSUE THREAD

**Issue by MichaelClerx** [MichaelClerx/myokit#727]
SubCapture is used to hide compiler output, and doesn't look like there's another way to suppress that...

Users might run myokit in threads (shared memory) or multiprocessing (no shared memory), so need to make sure this doesn't cause any issues

**Comment by MichaelClerx** [comment]
First test case:
```
#!/usr/bin/env python3
import myokit

x = myokit.SubCapture()
y = myokit.SubCapture()
z = myokit.SubCapture()

print('No capture')

x.enable()
print('x enabled')

y.enable()
print('y enabled')

z.enable()
print('z enabled')

y.disable()
print('y disabled')

x.disable()
print('x disabled')

z.disable()
print('z disabled')
```
Should only show 'No capture' and 'z disabled', but at the moment shows 'No capture', then 'x disabled', then nothing

**Comment by MichaelClerx** [comment]
## Test case 2
No shared memory, starting new processes with `spawn`
This is how windows does it, and should lead to less resource-sharing
```
#!/usr/bin/env python3
import multiprocessing
import os
import time

import myokit


def pr(x):
    print(str(os.getpid()) + ': ' + str(x))


def f(t=0):
    pr('Running with t=' + str(t))
    with myokit.SubCapture():
        pr('Capturing')
        if t:
            pr('Sleeping')
            time.sleep(t)
            pr('Waking')
    pr('No longer capturing')


if __name__ == '__main__':
    multiprocessing.set_start_method('spawn')

    xs = [1, 5, 10]
    ps = []
    for x in xs:
        p = multiprocessing.Process(target=f, args=(x,))
        p.start()
        ps.append(p)
        time.sleep(0.5)

    for p in ps:
        if p.is_alive():
            p.join()
```
Example output:
```
[michael@laptop1 myokit]$ ./spawn.py 
WARNING:myokit:Using development version of Myokit (1.32.1.dev).
WARNING:myokit:Using development version of Myokit (1.32.1.dev).
12773: Running with t=1
WARNING:myokit:Using development version of Myokit (1.32.1.dev).
12797: Running with t=5
WARNING:myokit:Using development version of Myokit (1.32.1.dev).
12803: Running with t=10
12773: No longer capturing
12797: No longer capturing
12803: No longer capturing
```
and
```
[michael@laptop1 myokit]$ ./spawn.py 
WARNING:myokit:Using development version of Myokit (1.32.1.dev).
WARNING:myokit:Using development version of Myokit (1.32.1.dev).
12860: Running with t=1
WARNING:myokit:Using development version of Myokit (1.32.1.dev).
12866: Running with t=5
WARNING:myokit:Using development version of Myokit (1.32.1.dev).
12860: No longer capturing
12872: Running with t=10
12866: No longer capturing
12872: No longer capturing
```
Seems that this case works fine!

**Comment by MichaelClerx** [comment]
## Test case 3
Multiprocessing (no shared memory) with `fork`
This is the linux default, and should share "many" things created before the forking occurs
```
#!/usr/bin/env python3
import multiprocessing
import os
import time

import myokit


def pr(x):
    print(str(os.getpid()) + ': ' + str(x))


def f(t=0):
    pr('Running with t=' + str(t))
    with myokit.SubCapture():
        pr('Capturing')
        if t:
            pr('Sleeping')
            time.sleep(t)
            pr('Waking')
    pr('No longer capturing')


if __name__ == '__main__':
    multiprocessing.set_start_method('fork')

    xs = [5, 1, 8]
    ps = []
    for x in xs:
        p = multiprocessing.Process(target=f, args=(x,))
        p.start()
        ps.append(p)
        time.sleep(0.3)

    for p in ps:
        if p.is_alive():
            p.join()
```
Example output:
```
[michael@laptop1 myokit]$ ./fork.py 
WARNING:myokit:Using development version of Myokit (1.32.1.dev).
13840: Running with t=5
13841: Running with t=1
13842: Running with t=8
13841: No longer capturing
13840: No longer capturing
13842: No longer capturing
```

**Comment by MichaelClerx** [comment]
Good news! Looks like it's only an issue with shared memory then

**Comment by MichaelClerx** [comment]
Closed in #728
